src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/gather_nonzero_ref.cl

```diff
@@ -5,83 +5,150 @@
 #include "include/batch_headers/common.cl"
 #include "include/batch_headers/data_types.cl"
 
-#define INPUT0_GET_INDEX1(idx_order) INPUT0_GET_INDEX(idx_order)
+#define VSIZE 8
+#define VLOAD CAT(vload, VSIZE)
+#define VSTORE CAT(vstore,VSIZE)
+#define OUTPUT_VTYPE MAKE_VECTOR_TYPE(OUTPUT_TYPE, VSIZE)
 
 KERNEL (gather_nonzero_ref)(const __global INPUT0_TYPE* input,
                             volatile __global INPUT1_TYPE* output_shape,
                             __global OUTPUT_TYPE* output)
 {
-    const uint gdim0 = (uint)get_global_id(0);
-    const uint gdim1 = (uint)get_global_id(1);
-    const uint gdim2 = (uint)get_global_id(2);
-
-    #if INPUT0_DIMS == 6
-        #define INPUT_ORDER b,f,w,z,y,x
-        const uint x = gdim0 % INPUT0_SIZE_X;
-        const uint y = gdim0 / INPUT0_SIZE_X;
-        const uint z = gdim1 % INPUT0_SIZE_Z;
-        const uint w = gdim1 / INPUT0_SIZE_Z;
-    #elif INPUT0_DIMS == 5
-        #define INPUT_ORDER b,f,z,y,x
-        const uint x = gdim0;
-        const uint y = gdim1 % INPUT0_SIZE_Y;
-        const uint z = gdim1 / INPUT0_SIZE_Y;
-    #elif INPUT0_DIMS == 4
-        #define INPUT_ORDER b,f,y,x
-        const uint x = gdim0;
-        const uint y = gdim1;
-    #endif
-
-    const uint f = gdim2 % INPUT0_FEATURE_NUM;
-    const uint b = gdim2 / INPUT0_FEATURE_NUM;
-
-    int num_nonzero_acc = (input[INPUT0_GET_INDEX1(INPUT_ORDER)] == INPUT0_VAL_ZERO) ? 0 : 1;
-    num_nonzero_acc = sub_group_scan_inclusive_add(num_nonzero_acc);
+    int local_offset = 0;
+    const int result_size = OV_INPUT_RANK * OUTPUT_FEATURE_NUM; // output shape: [ov_rank, count_nonzero]
+    const int max_local_mem_size = (64*1024)/(sizeof(OUTPUT_TYPE));
+
+#ifdef USE_LOCAL_MEM
+    __local OUTPUT_TYPE out_mem[max_local_mem_size];
+#else
+    __global OUTPUT_TYPE* out_mem = output;
+#endif
+
+    int b_inc, f_inc, y_inc, x_inc;
+    int count_nzero = output_shape[1];
+#if OV_INPUT_RANK == 1 // b
+    #define ADD_IDXS \
+        int b = input_idx_v; \
+        int b = input_idx_v / INPUT0_BATCH_PITCH; \
```

Comment:
It seems like that these two lines raise a redefinition error.

Reply:
Thnks, I fixed and added 1d,2d,3d tests too